src/lib/statistics/chunk_statistics/histograms/abstract_histogram.cpp

```diff
@@ -0,0 +1,620 @@
+        return total_count();
+      }
+
+      // This should never be false because can_prune should have been true further up if this was the case.
+      DebugAssert(value >= min(), "Value smaller than min of histogram.");
+
+      auto index = _bin_for_value(value);
+      auto cardinality = 0.f;
+
+      if (index == INVALID_BIN_ID) {
+        // The value is within the range of the histogram, but does not belong to a bin.
+        // Therefore, we need to sum up the counts of all bins with a max < value.
+        index = _upper_bound_for_value(value);
+      } else {
+        cardinality += _bin_share(index, value) * _bin_height(index);
+      }
+
+      // Sum up all bins before the bin (or gap) containing the value.
+      for (BinID bin = 0u; bin < index; bin++) {
+        cardinality += _bin_height(bin);
+      }
+
+      /**
+       * The cardinality is capped at total_count().
+       * It is possible for a value that is smaller than or equal to the max of the EqualHeightHistogram
+       * to yield a calculated cardinality higher than total_count.
+       * This is due to the way EqualHeightHistograms store the count for a bin,
+       * which is in a single value (count_per_bin) for all bins rather than a vector (one value for each bin).
+       * Consequently, this value is the desired count for all bins.
+       * In practice, _bin_count(n) >= _count_per_bin for n < bin_count() - 1,
+       * because bins are filled up until the count is at least _count_per_bin.
+       * The last bin typically has a count lower than _count_per_bin.
+       * Therefore, if we calculate the share of the last bin based on _count_per_bin
+       * we might end up with an estimate higher than total_count(), which is then capped.
+       */
+      return std::min(cardinality, static_cast<float>(total_count()));
+    }
+    case PredicateCondition::LessThanEquals:
+      return estimate_cardinality(PredicateCondition::LessThan, _get_next_value(value));
+    case PredicateCondition::GreaterThanEquals:
+      return total_count() - estimate_cardinality(PredicateCondition::LessThan, variant_value);
+    case PredicateCondition::GreaterThan:
+      return total_count() - estimate_cardinality(PredicateCondition::LessThanEquals, variant_value);
+    case PredicateCondition::Between: {
+      Assert(static_cast<bool>(variant_value2), "Between operator needs two values.");
+      const auto value2 = type_cast<T>(*variant_value2);
+
+      if (value2 < value) {
+        return 0.f;
+      }
+
+      return estimate_cardinality(PredicateCondition::LessThanEquals, *variant_value2) -
+             estimate_cardinality(PredicateCondition::LessThan, variant_value);
+    }
+    case PredicateCondition::Like:
+    case PredicateCondition::NotLike:
+      Fail("Predicate NOT LIKE is not supported for non-string columns.");
+    default:
+      // TODO(anyone): implement more meaningful things here
+      return total_count();
+  }
+}
+
+// Specialization for numbers.
```

Comment:
Why is this needed? Why can't we simply implement `estimate_cardinality` and specialize it for strings?

Reply:
This was the only way I could think of to not duplicate the code for the operators that work for both numbers and strings, i.e., all of them but `(NOT) LIKE`.
So the string specialization only deals with `(NOT) LIKE`, and for the others calls the same function as this specialization calls, which handles the universally supported operators.

Reply:
Is there a cleaner way to do that without duplicating code?